frontend/src/pages/SourceEdit.tsx

```diff
@@ -213,27 +384,71 @@ const SourceEdit: React.FC = () => {
       if (!originalSourceId || !translatedSourceId) {
         return;
       }
-      // Make sure all segments are saved?!
-      const segmentsToTranslate = [];
-      for (const originalSegment of segments[originalSourceId]) {
+
+      let originalSourceSegments = segments[originalSourceId];
+      if (!originalSourceSegments || originalSourceSegments.length === 0) {
+        originalSourceSegments = await dispatch(fetchSegments({ source_id: originalSourceId })).unwrap();
+      }
+
+      if (!originalSourceSegments || originalSourceSegments.length === 0) {
+        showToast('No original segments available for translation.', 'info');
+        return;
+      }
+
+      const segmentsToTranslate: Segment[] = [];
+      for (const originalSegment of originalSourceSegments) {
         if (!translatedSegmentsByOrder[originalSegment.order] && !translations[originalSegment.order]) {
           segmentsToTranslate.push(originalSegment);
         }
         if (segmentsToTranslate.length >= 10) {
           break;
         }
-      };
-      await translateSegments(segmentsToTranslate, translatedSourceId, originalLanguage, translatedLanguage);
+      }
+
+      if (!segmentsToTranslate.length) {
+        showToast('No pending segments to translate.', 'info');
+        return;
+      }
+
+      const additionalSourcesText = await getAdditionalSourcesRemainingText();
```

Comment:
You don't need to extract this, it is written in the database and will be used in backend.